lib/seahorn/Smt/TypeChecker.cc

```diff
@@ -3,44 +3,158 @@
 #include "seahorn/Expr/ExprLlvm.hh"
 #include "seahorn/Expr/ExprVisitor.hh"
 #include "seahorn/Support/SeaDebug.h"
+#include "seahorn/Support/SeaLog.hh"
 #include "llvm/Support/raw_ostream.h"
 
 using namespace expr;
-
 namespace {
 // local code in this namespace
-
 //==-- main implementation goes here --==//
 class TCVR {
 
+  ExprMap m_cache;
+
+  bool m_isWellFormed;
+  Expr m_errorExp;
+
+  // keeps track of every error's first/bottom-most sub expression
+  // for example, (bool && (int || int )) will map to (int || int)
+  ExprMap m_errorMap;
+
+  TypeCheckerHelper *const m_helper;
+
+  // Keeps track of the expression that the typechecker is called with.
+  // The expression is done traversing when it reaches the top most expression
+  // so it can reset
+  Expr m_topMost;
+
+  // Keeps track of all the bound variables that an expression uses
+  std::map<Expr, ExprSet> m_boundVarMap;
```

Comment:
as in the comment elsewhere, bound variables are more complex than they appear, let's ignore them for now